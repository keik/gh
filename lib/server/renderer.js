// @flow

/**
 * Return full HTML string with React props
 *
 * @param {string} content content as HTML string
 * @param {object} props React props object for Server-Side Rendering
 * @param {string} entryName name of entrypoint
 */
export function renderFullPage(
  content: string,
  props: any,
  entryName: string,
  bust: number | string
) {
  return `
 <!DOCTYPE html>
 <html>
   <head>
     <meta charset="utf-8">
     <title>gh</title>
     <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css">
     <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.7.0/styles/github.min.css">
     <link href="https://fonts.googleapis.com/css?family=Rajdhani" rel="stylesheet" />
     <link rel="stylesheet" href="/style.css?${bust}">
   </head>
   <body>
     <div id="app">${content}</div>
     <script>APP_PROPS = ${JSON.stringify(props)}</script>
     <script src="/${entryName}.js?${bust}"></script>
   </body>
 </html>
 `
}
