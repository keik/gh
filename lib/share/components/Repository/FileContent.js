// @flow

import { highlight, getLanguage } from 'highlight.js'
import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router'
import type { Dispatch } from 'redux'
import { extname } from 'path'

import Button from '../common/atoms/Button'
import Code from '../common/blocks/Code'
import Panel from '../common/blocks/Panel'
import InnerContainer from '../common/layouts/InnerContainer'
import SegmentedButtonsContainer from '../common/layouts/SegmentedButtonsContainer'
import type { ReducersStateT } from '../../ducks'
import * as EntriesAction from '../../ducks/repository/entries'
import type { EntryT } from '../../../types/gh'

export const FileContent = ({
  contributors = [],
  entry,
  params
}: {
  contributors: Array<string>,
  entry: ?EntryT,
  params: {
    branch: string,
    owner: string,
    repo: string,
    splat: string
  }
}) => {
  if (entry == null) return null

  const { content, size } = entry
  const lines = [].concat(content.match(/\n/g)).length
  const ext = extname(params.splat).substr(1)
  return (
    <InnerContainer>
      <Panel>
        <Panel.Header info>commit tease</Panel.Header>
        <Panel.Body id="contributors">
          <button>{contributors.length} contributors</button>
          {contributors.map(c => (
            <Link key={c} to={`/${c}`}>
              <img
                alt={c}
                style={{
                  verticalAlign: 'middle',
                  width: 20,
                  height: 20,
                  background: '#ccc'
                }}
              />
            </Link>
          ))}
        </Panel.Body>
      </Panel>
      <Panel>
        <Panel.Header>
          <div style={{ display: 'flex', justifyContent: 'space-between' }}>
            <div>
              {lines} line | {size} bytes
            </div>
            <div style={{ display: 'flex' }}>
              <SegmentedButtonsContainer>
                <Button
                  as={Link}
                  // $FlowFixMe
                  small="true"
                  to={`/{params.owner}/${params.repo}/raw/${params.branch}/${
                    params.splat
                  }`}
                >
                  Raw
                </Button>
                <Button
                  as={Link}
                  // $FlowFixMe
                  small="true"
                  to={`/${params.owner}/${params.repo}/blame/${params.branch}/${
                    params.splat
                  }`}
                >
                  Blame
                </Button>
                <Button
                  as={Link}
                  // $FlowFixMe
                  small="true"
                  to={`/${params.owner}/${params.repo}/commits/${
                    params.branch
                  }/${params.splat}`}
                >
                  History
                </Button>
              </SegmentedButtonsContainer>
              <button style={{ margin: '0 6px' }}>
                <i className="fa fa-pencil" />
              </button>
              <button>
                <i className="fa fa-trash" />
              </button>
            </div>
          </div>
        </Panel.Header>
        <Panel.Body>
          <Code>
            <Code.Lines>
              {Array.from(Array(lines).keys()).map(i => (
                <li key={i + 1}>{i + 1}</li>
              ))}
            </Code.Lines>
            <Code.Body
              className="hljs"
              dangerouslySetInnerHTML={{
                __html: getLanguage(ext)
                  ? highlight(ext, content).value
                  : content
              }}
              id="content"
            />
          </Code>
        </Panel.Body>
      </Panel>
    </InnerContainer>
  )
}

export default connect(
  ({ entries }: ReducersStateT, { params: { splat: path } }) => ({
    entry: entries.find(e => (e.path = path))
  })
)(
  class FileContentContainer extends React.Component<{
    contributors: Array<string>,
    dispatch: Dispatch<*>,
    entry: ?EntryT,
    params: {
      branch: string,
      owner: string,
      repo: string,
      splat: string
    }
  }> {
    async componentDidMount() {
      const {
        dispatch,
        params: { branch, owner, repo, splat: path = '' }
      } = this.props
      dispatch(await EntriesAction.fetch({ owner, repo, path, branch }))
    }

    render() {
      return <FileContent {...this.props} />
    }
  }
)
