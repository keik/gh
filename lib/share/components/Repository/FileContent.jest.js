// @flow

import * as React from 'react'
import renderer from 'react-test-renderer'

import { FileContent } from './FileContent'

test('<FileContent /> snapshot', () => {
  const tree = renderer
    .create(
      <FileContent
        contributors={['$CONTRIBUTOR_1', '$CONTRIBUTOR_2']}
        entry={{
          content: '$CONTENT',
          name: '$NAME',
          path: '$PATH',
          sha: '$SHA',
          size: 10,
          type: '$TYPE',
          url: '$URL'
        }}
        params={{
          branch: '$BRANCH',
          owner: '$OWNER',
          repo: '$REPO',
          splat: '$SPLAT'
        }}
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
