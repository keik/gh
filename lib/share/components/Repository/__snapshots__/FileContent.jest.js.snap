// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FileContent /> snapshot 1`] = `
<div
  className="InnerContainer-sc-1toru5t-0 eVVvlf"
>
  <div
    className="Panel-sc-1hnzj4y-0 ccxQvz"
  >
    <div
      className="Panel__Header-sc-1hnzj4y-1 gIbZlT"
    >
      commit tease
    </div>
    <div
      className="Panel__Body-sc-1hnzj4y-2 dwhEKI"
      id="contributors"
    >
      <button>
        2
         contributors
      </button>
      <a
        onClick={[Function]}
        style={Object {}}
      >
        <img
          alt="$CONTRIBUTOR_1"
          style={
            Object {
              "background": "#ccc",
              "height": 20,
              "verticalAlign": "middle",
              "width": 20,
            }
          }
        />
      </a>
      <a
        onClick={[Function]}
        style={Object {}}
      >
        <img
          alt="$CONTRIBUTOR_2"
          style={
            Object {
              "background": "#ccc",
              "height": 20,
              "verticalAlign": "middle",
              "width": 20,
            }
          }
        />
      </a>
    </div>
  </div>
  <div
    className="Panel-sc-1hnzj4y-0 ccxQvz"
  >
    <div
      className="Panel__Header-sc-1hnzj4y-1 chmgYi"
    >
      <div
        style={
          Object {
            "display": "flex",
            "justifyContent": "space-between",
          }
        }
      >
        <div>
          1
           line | 
          10
           bytes
        </div>
        <div
          style={
            Object {
              "display": "flex",
            }
          }
        >
          <div
            className="SegmentedButtonsContainer-sc-1alpooo-0 cmuWFA"
          >
            <a
              className="Button-sc-1ldg16x-0 gIWfju"
              onClick={[Function]}
              small="true"
              style={Object {}}
            >
              Raw
            </a>
            <a
              className="Button-sc-1ldg16x-0 gIWfju"
              onClick={[Function]}
              small="true"
              style={Object {}}
            >
              Blame
            </a>
            <a
              className="Button-sc-1ldg16x-0 gIWfju"
              onClick={[Function]}
              small="true"
              style={Object {}}
            >
              History
            </a>
          </div>
          <button
            style={
              Object {
                "margin": "0 6px",
              }
            }
          >
            <i
              className="fa fa-pencil"
            />
          </button>
          <button>
            <i
              className="fa fa-trash"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      className="Panel__Body-sc-1hnzj4y-2 dwhEKI"
    >
      <pre
        className="Code-nifwyh-0 fANLkT"
      >
        <ul
          className="Code__Lines-nifwyh-1 cHpdyK"
        >
          <li>
            1
          </li>
        </ul>
        <code
          className="hljs Code__Body-nifwyh-2 kloHoE"
          dangerouslySetInnerHTML={
            Object {
              "__html": "$CONTENT",
            }
          }
          id="content"
        />
      </pre>
    </div>
  </div>
</div>
`;
